src/foam/lib/parse/StringPStream.java

```diff
@@ -77,4 +78,12 @@ public String substring(PStream end) {
   public PStream apply(Parser ps, ParserContext x) {
     return ps.parse(this, x);
   }
+
+  public int decrement() {
+    return this.pos = this.pos == 0 ? 0 : this.pos-1;
+  }
+
+  public char beforeHead() {
+    return pos == 0 ?  head() :str.get().charAt(pos-1);
```

Comment:
Same thing here, you can backtrack by remembering the previous pstream object, I don't think we should need to add this.